File: Examples/ios/Klipper-Example/Klipper-Example/KMMView.swift
//
//  KMMView.swift
//  Klipper-Example
//
//  Created by Andrew Reed on 02/06/2023.
//

import Foundation
import SwiftUI
import library

struct KMMView: View {
    
    let viewModel = CatFactsViewModel(context: nil)
    
    var body: some View {
        VStack {
            Image(systemName: "globe")
                .imageScale(.large)
                .foregroundColor(.accentColor)
            Button(action: {
                viewModel.startClient()
            }, label: {
                Text("Connect to flipper")
            })
            Button(action: {
                viewModel.makeNetworkRequest()
            }, label: {
                Text("Make network request!")
            })
        }.onDisappear {
            viewModel.closeClient()
        }
        .padding()
    }
}

File: Examples/android/app/src/main/java/com/myunidays/klipperexample/KmmView.kt
package com.myunidays.klipperexample

import android.content.Context
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.Button
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ViewModel
import com.myunidays.library.CatFactsViewModel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class KMMViewModel(context: Context) : ViewModel() {
    private val viewModel = CatFactsViewModel(context = context)

    init {
        runCatching {
            viewModel.closeClient()
        }
    }

    fun startClient() = viewModel.startClient()
    fun makeNetworkRequest() {
        CoroutineScope(Dispatchers.IO).launch {
            viewModel.makeNetworkRequest()
        }
    }
}

@Composable
fun KmmView(viewModel: KMMViewModel = KMMViewModel(context = LocalContext.current)) {

    Column(
        modifier = Modifier.fillMaxWidth(),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Spacer(modifier = Modifier.weight(1f))
        Button(onClick = {
            viewModel.startClient()
        }) {
            Text(text = "Connect to flipper")
        }
        Button(onClick = {
            viewModel.makeNetworkRequest()
        }) {
            Text(text = "KMM Make network request")
        }
        Spacer(modifier = Modifier.weight(1f))
    }
}
